File: backend/src/atlas/schemas/section16.py
"""Pydantic schemas for Section 16 — Entry Gatekeeper.

Section 16 is the entry-decision pipeline.  It evaluates 4 rules per ticker
(Track A) or 2 rules + parabolic exception (Track B), plus an optional
override path (Track A only).  When the gate PASSES, control hands off to
Framework 12 for sizing.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Enums (string literals — kept simple for JSON serialisation).
# ---------------------------------------------------------------------------

TrackType = Literal["TRACK_A", "TRACK_B", "UNASSIGNED"]
GateResult = Literal["PASS", "FAIL", "UNKNOWN"]
Rule1Priority = Literal["PRIORITY_1", "PRIORITY_2", "PRIORITY_3", "NO_MATCH"]


# ---------------------------------------------------------------------------
# Per-rule result models — each rule returns its own structured verdict.
# ---------------------------------------------------------------------------


class Rule1Result(BaseModel):
    """Rule 1 — Catalyst conviction gate (priority-based signal threshold)."""

    result: GateResult
    priority_matched: Rule1Priority
    dark_pool_usd: float | None = None
    flow_usd: float | None = None
    days_to_earnings: int | None = None
    threshold_dark_pool_usd: float | None = None
    threshold_flow_usd: float | None = None
    reason: str


class Rule2Result(BaseModel):
    """Rule 2 — Catalyst horizon (must have earnings/catalyst within window)."""

    result: GateResult
    days_to_earnings: int | None = None
    earnings_date: date | None = None
    catalyst_max_days: int
    parabolic_window: bool = False
    reason: str


class Rule3Result(BaseModel):
    """Rule 3 — Price-position gate (not near 365d high without a pullback)."""

    result: GateResult
    current_price: float | None = None
    high_365d: float | None = None
    pct_below_high: float | None = None
    near_high_pct_threshold: float
    pullback_pct_required: float
    pullback_pct_actual: float | None = None
    reason: str


class Rule4Result(BaseModel):
    """Rule 4 — Portfolio fit (operator must set fits_portfolio=True today)."""

    result: GateResult
    fit_date: date | None = None
    fits_portfolio: bool | None = None
    set_by: str | None = None
    cluster_gap: str | None = None
    redundancy_check: str | None = None
    reason: str


class OverrideResult(BaseModel):
    """Override evaluation — Track A only.  Bypasses Rule 1 fail."""

    available: bool
    used_in_cycle: bool
    qualifies: bool
    dark_pool_usd: float | None = None
    flow_usd: float | None = None
    threshold_dark_pool_usd: float
    threshold_flow_usd: float
    lookback_days: int
    reason: str


# ---------------------------------------------------------------------------
# Top-level Section 16 result — what every consumer reads.
# ---------------------------------------------------------------------------


class Section16Result(BaseModel):
    """Entry-gate verdict for a single ticker.

    `gate` is the single source of truth: PASS means hand off to Framework 12.
    """

    ticker: str
    track: TrackType
    gate: GateResult

    rule1: Rule1Result | None = None
    rule2: Rule2Result | None = None
    rule3: Rule3Result | None = None
    rule4: Rule4Result | None = None
    override: OverrideResult | None = None

    # When true, the override (Track A) was used to satisfy Rule 1.
    override_used: bool = False

    evaluated_at: datetime
    notes: str | None = None


# ---------------------------------------------------------------------------
# Request bodies for operator-set state.
# ---------------------------------------------------------------------------


class TrackAssignmentRequest(BaseModel):
    """Body for POST /section16/track/{ticker}."""

    track: Literal["TRACK_A", "TRACK_B"]
    assigned_by: str = Field(min_length=1, max_length=100)
    notes: str | None = None


class Rule4Request(BaseModel):
    """Body for POST /section16/rule4/{ticker}."""

    fits_portfolio: bool
    set_by: str = Field(min_length=1, max_length=100)
    cluster_gap: str | None = None
    redundancy_check: str | None = None
    notes: str | None = None


class OverrideUseRequest(BaseModel):
    """Body for POST /section16/override/{ticker}/use."""

    used_by: str = Field(min_length=1, max_length=100)
    notes: str | None = None

File: frontend/src/lib/schemas/section16.ts
import { z } from 'zod';

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export const trackTypeSchema = z.enum(['TRACK_A', 'TRACK_B', 'UNASSIGNED']);
export type TrackType = z.infer<typeof trackTypeSchema>;

export const gateResultSchema = z.enum(['PASS', 'FAIL', 'UNKNOWN']);
export type GateResult = z.infer<typeof gateResultSchema>;

export const rule1PrioritySchema = z.enum([
  'PRIORITY_1',
  'PRIORITY_2',
  'PRIORITY_3',
  'NO_MATCH',
]);
export type Rule1Priority = z.infer<typeof rule1PrioritySchema>;

// ---------------------------------------------------------------------------
// Per-rule result schemas
// ---------------------------------------------------------------------------

export const rule1ResultSchema = z.object({
  result: gateResultSchema,
  priority_matched: rule1PrioritySchema,
  dark_pool_usd: z.number().nullable(),
  flow_usd: z.number().nullable(),
  days_to_earnings: z.number().nullable(),
  threshold_dark_pool_usd: z.number().nullable(),
  threshold_flow_usd: z.number().nullable(),
  reason: z.string(),
});
export type Rule1Result = z.infer<typeof rule1ResultSchema>;

export const rule2ResultSchema = z.object({
  result: gateResultSchema,
  days_to_earnings: z.number().nullable(),
  earnings_date: z.string().nullable(),
  catalyst_max_days: z.number(),
  parabolic_window: z.boolean(),
  reason: z.string(),
});
export type Rule2Result = z.infer<typeof rule2ResultSchema>;

export const rule3ResultSchema = z.object({
  result: gateResultSchema,
  current_price: z.number().nullable(),
  high_365d: z.number().nullable(),
  pct_below_high: z.number().nullable(),
  near_high_pct_threshold: z.number(),
  pullback_pct_required: z.number(),
  pullback_pct_actual: z.number().nullable(),
  reason: z.string(),
});
export type Rule3Result = z.infer<typeof rule3ResultSchema>;

export const rule4ResultSchema = z.object({
  result: gateResultSchema,
  fit_date: z.string().nullable(),
  fits_portfolio: z.boolean().nullable(),
  set_by: z.string().nullable(),
  cluster_gap: z.string().nullable(),
  redundancy_check: z.string().nullable(),
  reason: z.string(),
});
export type Rule4Result = z.infer<typeof rule4ResultSchema>;

export const overrideResultSchema = z.object({
  available: z.boolean(),
  used_in_cycle: z.boolean(),
  qualifies: z.boolean(),
  dark_pool_usd: z.number().nullable(),
  flow_usd: z.number().nullable(),
  threshold_dark_pool_usd: z.number(),
  threshold_flow_usd: z.number(),
  lookback_days: z.number(),
  reason: z.string(),
});
export type OverrideResult = z.infer<typeof overrideResultSchema>;

// ---------------------------------------------------------------------------
// Top-level result
// ---------------------------------------------------------------------------

export const section16ResultSchema = z.object({
  ticker: z.string(),
  track: trackTypeSchema,
  gate: gateResultSchema,
  rule1: rule1ResultSchema.nullable(),
  rule2: rule2ResultSchema.nullable(),
  rule3: rule3ResultSchema.nullable(),
  rule4: rule4ResultSchema.nullable(),
  override: overrideResultSchema.nullable(),
  override_used: z.boolean(),
  evaluated_at: z.string(),
  notes: z.string().nullable(),
});
export type Section16Result = z.infer<typeof section16ResultSchema>;

export const section16GateSchema = z.object({
  ticker: z.string(),
  track: trackTypeSchema,
  gate: gateResultSchema,
  override_used: z.boolean(),
});
export type Section16Gate = z.infer<typeof section16GateSchema>;

// ---------------------------------------------------------------------------
// Mutation request bodies
// ---------------------------------------------------------------------------

export const trackAssignmentRequestSchema = z.object({
  track: z.enum(['TRACK_A', 'TRACK_B']),
  assigned_by: z.string().min(1).max(100),
  notes: z.string().nullable().optional(),
});
export type TrackAssignmentRequest = z.infer<typeof trackAssignmentRequestSchema>;

export const trackAssignmentResponseSchema = z.object({
  ticker: z.string(),
  track: z.string(),
  assigned_by: z.string(),
  notes: z.string().nullable(),
});
export type TrackAssignmentResponse = z.infer<typeof trackAssignmentResponseSchema>;

export const rule4RequestSchema = z.object({
  fits_portfolio: z.boolean(),
  set_by: z.string().min(1).max(100),
  cluster_gap: z.string().nullable().optional(),
  redundancy_check: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),
});
export type Rule4Request = z.infer<typeof rule4RequestSchema>;

export const rule4ResponseSchema = z.object({
  ticker: z.string(),
  fit_date: z.string(),
  fits_portfolio: z.boolean(),
  set_by: z.string(),
});
export type Rule4Response = z.infer<typeof rule4ResponseSchema>;

export const overrideUseRequestSchema = z.object({
  used_by: z.string().min(1).max(100),
  notes: z.string().nullable().optional(),
});
export type OverrideUseRequest = z.infer<typeof overrideUseRequestSchema>;

export const overrideUseResponseSchema = z.object({
  ticker: z.string(),
  earnings_cycle_start: z.string(),
  earnings_cycle_end: z.string(),
  override_used: z.boolean(),
  override_used_by: z.string(),
});
export type OverrideUseResponse = z.infer<typeof overrideUseResponseSchema>;
